cmd/snap/cmd_run.go

```diff
@@ -216,6 +220,68 @@ func snapRunHook(snapName, snapRevision, hookName string) error {
 	return runSnapConfine(info, hook.SecurityTag(), snapName, "", hook.Name, nil)
 }
 
+func migrateXauthority(info *snap.Info) error {
+	u, err := userCurrent()
+	if err != nil {
+		return fmt.Errorf(i18n.G("cannot get the current user: %v"), err)
+	}
+
+	xauthPath := osGetenv("XAUTHORITY")
+	if len(xauthPath) == 0 || !osutil.FileExists(xauthPath) {
+		// Nothing to do for us. Most likely running outside of any
+		// graphical X11 session.
+		return nil
+	}
```

Comment:
I think the way to do this is change the algorithm a bit. This also means we don't need a tempfile which is something Gustavo was looking for in the forum

1. if defined and not zero length, open XAUTHORITY to get a file object and return nil if it doesn't exist
2. perform all the above recommended checks on file.Name() (the name that was used to open the file) and return nil in case of shenanigans or not in /tmp
3. read the contents from the file object
4. if the file contents are the same as what is already in XDG_RUNTIME_DIR/.Xauthority, set XAUTHORITY (via snapenv?) and return
5. otherwise, validate the content of the file object. if invalid, return nil with warning
6. since valid, securely write it out to XDG_RUNTIME_DIR/.Xauthority, 0600 and owned by userCurrent (let's just set up the permissions securely and not worry about copying the perms). set XAUTHORITY (via snapenv?) and return